source/Timer.cpp

```diff
@@ -0,0 +1,253 @@
+/* Timer.cpp
+Copyright (c) 2023 by Timothy Collett
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Timer.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+#include "Ship.h"
+#include "StellarObject.h"
+#include "UI.h"
+
+using namespace std;
+
+Timer::Timer(const DataNode &node, Mission *mission)
+{
+	Load(node, mission);
+}
+
+void Timer::Load(const DataNode &node, Mission *mission)
+{
+	if(node.Size() > 1)
+		name = node.Token(1);
+	if(node.Size() > 2)
+		base = static_cast<int64_t>(node.Value(2));
+	if(node.Size() > 3)
+		rand = static_cast<uint32_t>(node.Value(3));
+
+	this->mission = mission;
+
+	timeToWait = base + Random::Int(rand);
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "idle")
+			requireIdle = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "systems")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity")
+		{
+			if(child.Size() > 1)
+				proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+			if(child.HasChildren())
+			{
+				const DataNode &firstGrand = (*child.begin());
+				proximityCenter = GameData::Planets().Get(firstGrand.Token(0));
+			}
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset")
+		{
+			if(child.Size() > 1 && child.Token(1) == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(child.Size() > 1 && child.Token(1) == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(child.Size() > 1 && child.Token(1) == "none")
+				resetCondition = ResetCondition::NONE;
+			else if(child.Size() > 1 && child.Token(1) == "pause")
+				resetCondition = ResetCondition::PAUSE;
+		}
+		// We keep "on timeup" as separate tokens so that it's compatible with MissionAction syntax
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
```

Comment:
`on complete`?

Reply:
I chose `on timeup` so that it wouldn't cause confusion with the mission-level `on complete` trigger; I think that it both makes it a little easier for content creators to keep track of, and potentially allows for better future-proofing, as AIUI it gives the MissionAction object a distinct `trigger`.